<img src="./unal.png" align=left width="150" height="150"></img>

$\;$

---
<h2 align="center">Test de razón de verosimilitudes,  Wald, score, gradiente</h2>

### Oscar Julian Layton


A continuación se muestra un conjunto de datos que representan el número de roturas de urdimbre por telar, donde un telar corresponde a una longitud fija de hilo. Donde:

* **Breaks:** Número de roturas.
* **Wool:** Tipo de lana (A,B).
* **Tension:** Nivel de tensión (L,M,H).

In [2]:
data(warpbreaks)
head(warpbreaks,n = 10)

breaks wool tension
1  26     A    L      
2  30     A    L      
3  54     A    L      
4  25     A    L      
5  70     A    L      
6  52     A    L      
7  51     A    L      
8  26     A    L      
9  67     A    L      
10 18     A    M

Al realizarse una regresión se quiere ver si el número de roturas en que cada uno de los dispositivos funcionó en el modo 1 po en el modo 2 me ayuda a explicar las diferencias en el numero de fallas que rpesentaron los disposiitivos. Por consiguiente se procede a hacer una regresión con respuesta Poisson, $Y_k$ es Poisson con media $\mu_k$ donde $\mu_k$ depende de las covariables (intercepto $\beta_1$, $\beta_2$ que depende de wool,  $\beta_3$ que depende de tension), a qui la función de enlace es la canonica que en caso de la Poisson en la **LOGARITMO**.

In [3]:
fit <- glm(breaks ~ wool + tension, family = poisson(link="log"), data=warpbreaks) 
summary(fit)


Call:
glm(formula = breaks ~ wool + tension, family = poisson(link = "log"), 
    data = warpbreaks)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-3.6871  -1.6503  -0.4269   1.1902   4.2616  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  3.69196    0.04541  81.302  < 2e-16 ***
woolB       -0.20599    0.05157  -3.994 6.49e-05 ***
tensionM    -0.32132    0.06027  -5.332 9.73e-08 ***
tensionH    -0.51849    0.06396  -8.107 5.21e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance: 297.37  on 53  degrees of freedom
Residual deviance: 210.39  on 50  degrees of freedom
AIC: 493.06

Number of Fisher Scoring iterations: 4


Para evaluar si:

$$H_{0}: \beta_2 = \beta_3=\beta_4 \;\;\;\;\; vs \;\;\;\;\;\;H_{1}: \beta_2 \neq \beta_2 \neq \beta_3 \neq  0 $$

Considerando que el test de verosimilitudes está determinado por: $\xi _{RV}=2[l(\hat{\beta})-l(\hat{\beta^{o}})]$ sumando y restando $\hat{\beta}$ del modelo saturado se obtiene:

$$\xi _{RV}=2[l(\hat{\beta})-l(\hat{\beta^{o}})]=\phi^{-1} [D(y;\hat{\mu}^0)-D(y;\hat{\mu})]  $$

Donde $D(y;\hat{\mu}^0)$ es el desvio del modelo restringido llamado **Null Deviance** y donde $D(y;\hat{\mu})$ es el desvio del modelo sin restricción llamado **Residual Deviance**

**Null Deviance** Es el desvio de un modelo donde todos los betas excepto el intercepto son iguales a cero.


Luego como $\phi=1$ se tiene:

$$\xi _{RV}=Null deviance- Residual Deviance=297.37-210.39 =86.980$$


Luego se rechaza $H_0$ a un nivel de significancia APROXIMADO del 100$ \alpha\%$ puesto que $\xi _{RV}>\chi^2(1)$ A firmandose que essignificativo en el modelo.

In [4]:
(fit$null.deviance -fit$deviance)/1

[1] 86.98032

#### Segunda forma 

In [5]:
#test de razon de verosimilitudes
fit1<-glm(breaks~1 ,family = poisson(link="log"), data=warpbreaks)      # modelo solo ocn el intercepto
anova(fit1,fit, test = "LRT")                                           #Test razon de verosimilitudes

Resid. Df Resid. Dev Df Deviance Pr(>Chi)    
1 53        297.3722   NA       NA           NA
2 50        210.3919    3 86.98032 9.750414e-19

$$H_{0}: \beta_2 = \beta_3= \beta_4=0\;\;\;\;\; vs \;\;\;\;\;\; H_{1}: \beta_2 \neq \beta_3 \neq \beta_4 \neq 0 $$

Luego se rechaza $H_0$ a un nivel de significancia APROXIMADO del 100$ \alpha\%$ puesto que  ...........
estableciendo que almenos uno de los betas es significativo

#### Tercera forma 

In [6]:
#test de razon de verosimilitudes
fit2<-glm(breaks~1+wool ,family = poisson(link="log"), data=warpbreaks)      # modelo solo ocn el intercepto
anova(fit2,fit, test = "LRT")                                           #Test razon de verosimilitudes

Resid. Df Resid. Dev Df Deviance Pr(>Chi)    
1 52        281.3335   NA       NA           NA
2 50        210.3919    2 70.94157 3.937619e-16

$$H_{0}: \beta_3= \beta_4=0\;\;\;\;\; vs \;\;\;\;\;\; H_{1}: \beta_3 \neq \beta_4 \neq 0 $$

Luego se rechaza $H_0$ a un nivel de significancia APROXIMADO del 100$ \alpha\%$ estableciendo que la tensión es significativa en el numero de roturas.

### EJEMPLO DE HIPOTESIS

Se quiere verificar que la variable tensión afecta la media de la poisson (la variable tensión es significativa??), al observar la matriz diseño se puede ver que $\beta_3$ y $\beta_4$ son nuestro objetivo:

In [8]:
X <- model.matrix(fit);  head(X,n=14)                                      #Matriz diseño

(Intercept) woolB tensionM tensionH
1  1           0     0        0       
2  1           0     0        0       
3  1           0     0        0       
4  1           0     0        0       
5  1           0     0        0       
6  1           0     0        0       
7  1           0     0        0       
8  1           0     0        0       
9  1           0     0        0       
10 1           0     1        0       
11 1           0     1        0       
12 1           0     1        0       
13 1           0     1        0       
14 1           0     1        0

En este orden de ideas, si $\beta_3= \beta_4=0$ se puede decir que tensión no aporta al modelo. En cambio si alguno es diferente, tensión si aporta. En consecuencia la hipótesisi a considerar es:
$$H_{0}: \beta = 0 \;\;\;\;\; vs \;\;\;\;\;\; H_{1}: \beta \neq  0 $$

Donde $\beta$ (que restringimos) es de dimensión 2 donde $\beta = [\beta'_3 , \beta'_4]'$

Luego se tiene 4 elecciones de la estadística de prueba:

* Test Razón de verosimilitudes.
* Test de Wald.
* Test de Score.
* Test gradiente.

Estos gradientes pueden son contrastados con una $\chi^{2}_{2}$ con 2 DF

$$H_{0}: \beta_2 = \beta_3=0 \;\;\;\;\; vs \;\;\;\;\;\; H_{1}: \beta_2 \neq \beta_3 \neq 0 $$

In [9]:
#ESTAS LINEAS ES PARA LLAMAR LAS FUNCIONES REALIZADAS POR EL DOCENTE hernando vanegas
setwd("") #busca por default en una carpeta en especifica todo.
source("macros.txt")#Lllamar el archivo de macros que está en la carpeta especifica,(son una fuciines especificas del profesor)


In [10]:
#test de razon de verosimilitudes
fit1<-glm(breaks~1+ wool ,family = poisson(link="log"), data=warpbreaks)      # modelo solo ocn el intercepto
anova_glm(fit1,fit, test = "LRT")                                           #Test razon de verosimilitudes
1-pchisq(70.942,df = 2)         #pvalor

anova_glm(fit,fit1,test="wald") #test de wald
1-pchisq(71.051,df = 2)         #pvalor

anova_glm(fit,fit1,test="rao") #test de rao
1-pchisq(72.27,df = 2)         #pvalor

anova_glm(fit,fit1,test="gradient") #test de wald
1-pchisq(70.666,df = 2)         #pvalor


  Likelihood-ratio test 

 Model 1:  breaks ~ wool + tension 
 Model 2:  breaks ~ 1 + wool 

   Chi   Df  Pr(>Chi)    
  70.942  2 4.441e-16 ***


[1] 4.440892e-16


  Wald test 

 Model 1:  breaks ~ wool + tension 
 Model 2:  breaks ~ 1 + wool 

   Chi   Df  Pr(>Chi)    
  71.051  2 3.331e-16 ***


[1] 3.330669e-16


  Rao test 

 Model 1:  breaks ~ wool + tension 
 Model 2:  breaks ~ 1 + wool 

  Chi   Df Pr(>Chi)    
  72.27  2 2.22e-16 ***


[1] 2.220446e-16


  Gradient test 

 Model 1:  breaks ~ wool + tension 
 Model 2:  breaks ~ 1 + wool 

   Chi   Df  Pr(>Chi)    
  70.666  2 4.441e-16 ***


[1] 4.440892e-16

En todos los test se rechaza $H_0$

**Conclusión:** o $\beta_3$ o $\beta_4$ es diferente de cero, lo que quiere decir que tensión es significativo pues uno de los dos es significativo. En otras palabras tensión tiene un efecto significativo sobre la media de Y

In [11]:
t(coef(fit)[c(3,4)])%*%(vcov(fit)[c(3,4), c(3,4)])%*%(coef(fit)[c(3,4)])       #test de wald

[,1]       
[1,] 0.001983426

 
 ##  Nueva  hipótesis
 
 Ahora para hacer una hipótesis sobre el tipo de lana, considero a $\beta_2$. 

$$H_{0}: \beta_2 = 0 \;\;\;\;\; vs \;\;\;\;\;\; H_{1}: \beta_2 \neq  0 $$

donde 

$$ \xi _{W} = (\hat{\beta_2}  - \beta_2^{0})' [\hat{Var}(\hat{\beta_2})]{-1}(\hat{\beta_2}  - \beta_2^{0})$$

In [7]:
(coef(fit)[c(2)])^2 /(vcov(fit)[c(2), c(2)])     #test de wald

woolB 
15.95408

Aqui hay que notar que $\xi _{W}=15.95407$ es el mismo test de wald para el casi univariado, vease la salida

In [8]:
summary(fit)


Call:
glm(formula = breaks ~ wool + tension, family = poisson(link = "log"), 
    data = warpbreaks)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-3.6871  -1.6503  -0.4269   1.1902   4.2616  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  3.69196    0.04541  81.302  < 2e-16 ***
woolB       -0.20599    0.05157  -3.994 6.49e-05 ***
tensionM    -0.32132    0.06027  -5.332 9.73e-08 ***
tensionH    -0.51849    0.06396  -8.107 5.21e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance: 297.37  on 53  degrees of freedom
Residual deviance: 210.39  on 50  degrees of freedom
AIC: 493.06

Number of Fisher Scoring iterations: 4


Dese cuenta que $\xi _{W}=15.95407$ es el mismo valor de   $(zvalue)^2=(-3.994)^2 = 15.95407$ este valor es el que hay que comparar con una $\chi ^2 $ con un grado de libertad, 1df